src/datasets/formatting/formatting.py

```diff
@@ -0,0 +1,395 @@
+    if not isinstance(key, (int, slice, range, str, Iterable)):
+        _raise_bad_key_type(key)
+    if isinstance(key, str):
+        _check_valid_column_key(key, pa_table.column_names)
+    else:
+        size = len(indices) if indices is not None else pa_table.num_rows
+        _check_valid_index_key(key, size)
+    # Query the main table
+    if indices is None:
+        pa_subtable = _query_table(pa_table, key)
+    else:
+        pa_subtable = _query_table_with_indices_mapping(pa_table, key, indices=indices)
+    return pa_subtable
+
+
+def format_table(
+    pa_table: pa.Table,
+    key: Union[int, slice, range, str, Iterable],
+    formatter: Formatter,
+    format_columns: Optional[list] = None,
+    output_all_columns=False,
+):
+    """
+    Format a pyarrow Table depending on the key that was used and a Formatter object.
+
+    Args:
+        pa_table (``pyarrow.Table``): The input pyarrow Table to format
+        key (``Union[int, slice, range, str, Iterable]``): Depending on the key that was used, the formatter formats
+            the table as either a row, a column or a batch.
+        formatter (``datasets.formatting.formatting.Formatter``): Any subclass of a Formatter such as
+            PythonFormatter, NumpyFormatter, etc.
+        format_columns (Optional ``List[str]``): if not None, it defines the columns that will be formatted using the
+            given formatter. Other columns are discarded (unless ``output_all_columns`` is True)
+        output_all_columns (``bool``, defaults to False). If True, the formatted output is completed using the columns
+            that are not in the ``format_columns`` list. For these columns, the PythonFormatter is used.
+
+
+    Returns:
+        A row, column or batch formatted object defined by the Formatter:
+        - the PythonFormatter returns a dictionary for a row or a batch, and a list for a column.
+        - the NumpyFormatter returns a dictionary for a row or a batch, and a np.array for a column.
+        - the PandasFormatter returns a pd.DataFrame for a row or a batch, and a pd.Series for a column.
+        - the TorchFormatter returns a dictionary for a row or a batch, and a torch.Tensor for a column.
+        - the TFFormatter returns a dictionary for a row or a batch, and a tf.Tensor for a column.
+    """
+    query_type = key_to_query_type(key)
+    python_formatter = PythonFormatter()
+    if format_columns is None:
+        return formatter(pa_table, query_type=query_type)
+    elif query_type == "column":
+        if key in format_columns:
+            return formatter(pa_table, query_type)
+        else:
+            return python_formatter(pa_table, query_type=query_type)
+    else:
+        pa_table_to_format = pa_table.drop(col for col in pa_table.column_names if col not in format_columns)
+        formatted_output = formatter(pa_table_to_format, query_type=query_type)
+        if output_all_columns:
+            if hasattr(formatted_output, "update"):
```

Comment:
This is because the output of tokenizers (BatchEncoding) are not subclasses of dict, even if they implement parts of the dict interface (getitem, len, update etc.)

Although I just noticed that there actually `Mapping` objects so I can change to using `isinstance(formatted_output, Mapping)` which is better imo